README.md
# KindleDXPush

[![Latest Version][1]][2]
[![The MIT License][3]][4]

This project was created by [@blahgeek][5], now maintained by [@lord63][6].

## Intro

As we all know, we can use Amazon's free 3G network to deliver our docs, but we need
to manually click the deliver buttom in the browser. This script is born to rescue
us from this boring thing. Config this script, add it to the `crontab` and you'are done.
All your docs will be sent to your kindle automatically.


## Requirement

* Python 2.7
* Requests lib
* BeautifulSoup lib
* Linux platform(I haven't test it on windows)

## Install

    $ sudo pip install kindlepush

## Feature

* It has a log file, you can check which file you've delivered.

* Use sqlite database, don't worry about that a doc will be missed or delivered twice.

## Usage

    Usage: kindlepush <command> [option]
    Options:
      -h, --help            output the help menu
      -V, --version         output the version number
      -c, --count [count]   the count of the docs to deliver
      -n, --number [number] show how may logs
    Commands:
      read                  read the log file

first, touch a new file named `kindlepush_config.json` under `/usr/local/bin`, including those:

    {
        "email": "xxxxxx",              # your email
        "password": "xxxxxx",           # your amazon's password
        "directory": "/path/to/save/",  # save log file and database, end with '/'
        "count": 15,                    # check how many docs evert time that whether those have been deliverred before, default 15 is one page a time.
        "number": 4                     # the default count of log messages when you read from log file
    }

deliver your doc from your kindle library to your kindle:

    $ kindlepush
    Login...
    Delivering...
    delivering YOUR DOC
    Done. Save to db.
    delivering YOUR DOC
    Done. Save to db.

read the log file to get to know the docs which you have delivered:
(default is 4 messages, you can use `-n NUMBER` to get more information.)

    $ kindlepush read
    2014-09-23 07:40:32,077 [INFO] delivered YOURDOC
    2014-09-23 07:40:33,125 [INFO] delivered YOURDOC
    2014-09-23 07:40:34,183 [INFO] delivered YOURDOC
    2014-09-24 14:12:34,506 [INFO] delivered YOURDOC

Get help via `kindlepush -h` and `kindlepush read -h`.

## Wiki

see [wiki](https://github.com/lord63/kindledxpush/wiki)

## License

MIT

[1]: http://img.shields.io/pypi/v/kindlepush.svg
[2]: https://pypi.python.org/pypi/kindlepush
[3]: http://img.shields.io/badge/license-MIT-yellow.svg
[4]: https://github.com/lord63/kindledxpush/LICENSE
[5]: https://github.com/blahgeek
[6]: https://github.com/lord63


ChangeLog
2014.12.24
- v0.3.0 released
- add feature: get pending deliveries, close #5

2014.9.24
- v0.2.0 released
- add parameter "count" and version
- refine code style, logging, prompt, exception

2014.8.4:
- bring it to PyPI: v0.1.0 released
- get log information using parameter in the terminal

2014.7.20：
- automatically get deviceID

2014.5.9：
- diable the unwanted Requests log information

2014.4.24：
- refine code style, add comments
- add wiki

2014.4.23：
- record `title` rather than `asin` in log files.
- support chinese in log files.


main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import logging
import sqlite3
import sys
import re
import os
import json
from os import path

import requests
from bs4 import BeautifulSoup
from terminal import Command
from config import EMAIL, PASSWORD, COUNT

# Set default encoding from ascii to utf-8
reload(sys)
if hasattr(sys, 'setdefaultencoding'):
    sys.setdefaultencoding('utf-8')

# Connect to the database which lies in the current directory
db = sqlite3.connect(path.join(sys.path[0], 'main.db'))
cursor = db.cursor()

# User-agent is necessary, or it'll occur `KeyError: 'data' `.
s = requests.session()
s.headers.update({'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; \
                   rv:28.0) Gecko/20100101 Firefox/28.0'})


def translate(raw_title):
    '''translate escape characters to utf-8

    To support chinese in log.'''
    def trans(unicode):
        return unichr(int(unicode, 16)).encode('utf-8')

    if '&#x' not in raw_title:
        return raw_title
    characters = re.findall(r'(?<=&#x).{4}', raw_title)
    date = raw_title.split(';')[-1]

    title = ''
    for character in characters:
        title = title + trans(character)
    return title + date   # Date is a configuration in my KindleEar


def get_hidden_form_data(page):
    '''Get hidden form data from the page after you click the signin button

    Those data are needed to post to sign in.'''
    hidden_form_data = dict()
    input_tags = page.findAll('input', type='hidden')
    for tag in input_tags:
        hidden_form_data[tag['name']] = tag['value']
    return hidden_form_data


def login(email, password):
    '''To login in to amazom.com

    The long url is the 'sign in' botton, the short is the 'sign in our
    secure sever' botton.'''

    logging.info('Login...')
    print 'Login...'
    login_url = ('https://www.amazon.com/ap/signin?_encoding=UTF8&openid.assoc'
                 '_handle=usflex&openid.claimed_id=http%3A%2F%2Fspecs.openid.n'
                 'et%2Fauth%2F2.0%2Fidentifier_select&openid.identity=http%3A%'
                 '2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openi'
                 'd.mode=checkid_setup&openid.ns=http%3A%2F%2Fspecs.openid.net'
                 '%2Fauth%2F2.0&openid.ns.pape=http%3A%2F%2Fspecs.openid.net%2'
                 'Fextensions%2Fpape%2F1.0&openid.pape.max_auth_age=0&openid.r'
                 'eturn_to=https%3A%2F%2Fwww.amazon.com%2Fgp%2Fyourstore%2Fhom'
                 'e%3Fie%3DUTF8%26ref_%3Dgno_signin')
    login_post_url = 'https://www.amazon.com/ap/signin'
    login_page = BeautifulSoup(s.get(login_url).text)
    data = get_hidden_form_data(login_page)
    data.update({'email': email, 'password': password})
    s.post(login_post_url, data)


def get_contents():
    url = ('https://www.amazon.com/gp/digital/fiona/manage/features/'
           'order-history/ajax/queryPdocs.html')
    r = s.post(url, {'offset': 0, 'count': COUNT,
                     'contentType': 'Personal Document',
                     'queryToken': 0, 'isAjax': 1})
    return [{'category': item['category'], 'contentName': item['asin'],
            'title': item['title']} for item in r.json()['data']['items']]


def get_deviceID():
    r = s.post('https://www.amazon.com/mn/dcw/myx/ajax-activity',
              data={'data' :'{"param":{"GetDevices":{}}}'})
    deviceID = r.json()['GetDevices']['devices'][0]['deviceAccountId']
    return deviceID


def deliver_content(content):
    url = ('https://www.amazon.com/gp/digital/fiona/content-download/'
           'fiona-ajax.html/ref=kinw_myk_ro_send')
    content.update({'isAjax': '1', 'deviceID': get_deviceID()})
    r = s.post(url, content)
    assert r.json()['data'] == 1    # Whether successfully delivered it or not


def deliver_all(contents):
    '''Deliver all the contents which haven't been delivered before

    Once delivered it, save it's asin in database.'''
    logging.info('Delivering...')
    print 'Delivering...'

    def contentInDB(content):
        try:
            cursor.execute('select * from content where name = "%s"' % content)
        except sqlite3.OperationalError:
            cursor.execute('create table content (name text)')
            return False
        else:
            return False if cursor.fetchone() is None else True

    contents = filter(lambda x: not contentInDB(x['contentName']), contents)
    for content in contents:
        try:
            logging.info('delivering ' + translate(content['title']))
            print 'delivering ' + translate(content['title'])
            deliver_content(content)
        except:
            logging.error('Error, ignore')
            print 'Error, ignore'
            pass
        else:
            logging.info('Done. Save to db.')
            print 'Done. Save to db.'
            cursor.execute('insert into content values ("%s")' %
                           content['contentName'])
    db.commit()

def main():
    command = Command('kindledxpush', 'automatically deliver you doc to your kindle')

    try:
        with open(os.path.join(sys.path[0], 'kindle_config.json')) as f:
            config = json.load(f)
    except IOError:
        sys.exit("Check your config file please.")

    @command.action
    def read(number=config['number']):
        """
        read the log file

        :param number: the count of days
        :option number: -n, --number [number]
        """
        file_path = os.path.join(config['log_directory'], 'kindle.log')
        os.system('tail -n {0} {1}'.format(number, file_path))
    command.parse()

    logging.basicConfig(
            filename=os.path.join(config['log_directory'], 'kindle.log'),
            level='INFO',
            format='%(asctime)s [%(levelname)s] %(message)s')
    # Disable unwanted log message from the requests library
    requests_log = logging.getLogger("requests")
    requests_log.setLevel(logging.WARNING)

if __name__ == '__main__':
    login(EMAIL, PASSWORD)
    deliver_all(get_contents())


kindleush.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import logging
import sqlite3
import sys
import re
import os
import json

import requests
from bs4 import BeautifulSoup
from terminal import Command
#from config import EMAIL, PASSWORD, COUNT

# Set default encoding from ascii to utf-8
reload(sys)
if hasattr(sys, 'setdefaultencoding'):
    sys.setdefaultencoding('utf-8')

# User-agent is necessary, or it'll occur `KeyError: 'data' `.
s = requests.session()
s.headers.update({'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; \
                   rv:28.0) Gecko/20100101 Firefox/28.0'})


def translate(raw_title):
    '''translate escape characters to utf-8

    To support chinese in log.'''
    def trans(unicode):
        return unichr(int(unicode, 16)).encode('utf-8')

    if '&#x' not in raw_title:
        return raw_title
    characters = re.findall(r'(?<=&#x).{4}', raw_title)
    date = raw_title.split(';')[-1]

    title = ''
    for character in characters:
        title = title + trans(character)
    return title + date


def get_hidden_form_data(page):
    '''Get hidden form data from the page after you click the signin button

    Those data are needed to post to sign in.'''
    hidden_form_data = dict()
    input_tags = page.findAll('input', type='hidden')
    for tag in input_tags:
        hidden_form_data[tag['name']] = tag['value']
    return hidden_form_data


def login(config):
    '''To login in to amazom.com

    The long url is the 'sign in' botton, the short is the 'sign in our
    secure sever' botton.'''

    logging.info('Login...')
    print 'Login...'
    login_url = ('https://www.amazon.com/ap/signin?_encoding=UTF8&openid.assoc'
                 '_handle=usflex&openid.claimed_id=http%3A%2F%2Fspecs.openid.n'
                 'et%2Fauth%2F2.0%2Fidentifier_select&openid.identity=http%3A%'
                 '2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openi'
                 'd.mode=checkid_setup&openid.ns=http%3A%2F%2Fspecs.openid.net'
                 '%2Fauth%2F2.0&openid.ns.pape=http%3A%2F%2Fspecs.openid.net%2'
                 'Fextensions%2Fpape%2F1.0&openid.pape.max_auth_age=0&openid.r'
                 'eturn_to=https%3A%2F%2Fwww.amazon.com%2Fgp%2Fyourstore%2Fhom'
                 'e%3Fie%3DUTF8%26ref_%3Dgno_signin')
    login_post_url = 'https://www.amazon.com/ap/signin'
    login_page = BeautifulSoup(s.get(login_url).text)
    data = get_hidden_form_data(login_page)
    data.update({'email': config['email'], 'password': config['password']})
    s.post(login_post_url, data)


def get_contents(config):
    url = ('https://www.amazon.com/gp/digital/fiona/manage/features/'
           'order-history/ajax/queryPdocs.html')
    r = s.post(url, {'offset': 0, 'count': config['count'],
                     'contentType': 'Personal Document',
                     'queryToken': 0, 'isAjax': 1})
    return [{'category': item['category'], 'contentName': item['asin'],
            'title': item['title']} for item in r.json()['data']['items']]


def get_deviceID():
    r = s.post('https://www.amazon.com/mn/dcw/myx/ajax-activity',
              data={'data' :'{"param":{"GetDevices":{}}}'})
    deviceID = r.json()['GetDevices']['devices'][0]['deviceAccountId']
    return deviceID


def deliver_content(content):
    url = ('https://www.amazon.com/gp/digital/fiona/content-download/'
           'fiona-ajax.html/ref=kinw_myk_ro_send')
    content.update({'isAjax': '1', 'deviceID': get_deviceID()})
    r = s.post(url, content)
    assert r.json()['data'] == 1    # Whether successfully delivered it or not


def deliver_all(contents, db):
    '''Deliver all the contents which haven't been delivered before

    Once delivered it, save it's asin in database.'''
    logging.info('Delivering...')
    print 'Delivering...'

    def contentInDB(content):
        try:
            db.cursor.execute('select * from content where name = "%s"' % content)
        except sqlite3.OperationalError:
            db.cursor.execute('create table content (name text)')
            return False
        else:
            return False if db.cursor.fetchone() is None else True

    contents = filter(lambda x: not contentInDB(x['contentName']), contents)
    for content in contents:
        try:
            logging.info('delivering ' + translate(content['title']))
            print 'delivering ' + translate(content['title'])
            deliver_content(content)
        except:
            logging.error('Error, ignore')
            print 'Error, ignore'
            pass
        else:
            logging.info('Done. Save to db.')
            print 'Done. Save to db.'
            db.cursor.execute('insert into content values ("%s")' %
                           content['contentName'])
    db.commit()

def main():
    command = Command('kindledxpush', 'automatically deliver you doc to your kindle')

    try:
        with open(os.path.join(sys.path[0], 'kindle_config.json')) as f:
            config = json.load(f)
    except IOError:
        sys.exit("Check your config file please.")

    @command.action
    def read(number=config['number']):
        """
        read the log file

        :param number: the count of days
        :option number: -n, --number [number]
        """
        file_path = os.path.join(config['directory'], 'kindle.log')
        os.system('tail -n {0} {1}'.format(number, file_path))
    command.parse()

    logging.basicConfig(
            filename=os.path.join(config['directory'], 'kindle.log'),
            level='INFO',
            format='%(asctime)s [%(levelname)s] %(message)s')
    # Disable unwanted log message from the requests library
    requests_log = logging.getLogger("requests")
    requests_log.setLevel(logging.WARNING)

   # Connect to the database which lies in the current directory
    #db = sqlite3.connect(path.join(sys.path[0], 'main.db'))
    db = sqlite3.connect(os.path.join(config['directory'], 'main.db'))

    login(config)
    deliver_all(get_contents(config), db)

if __name__ == '__main__':
    main()


kindlepush/kindlepush.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    kindlepush
    ~~~~~~~~~~

    Kindlepush is trying to rescue you from manually clicking the deliver
    buttom to send the doc from your kindle library to your kindle. It
    is for 3G devices, such as kindle dx.
    It was created by @blahgeek, now maintained by @lord63.

    :copyright: (c) 2014 BlahGeek.
    :copyright: (c) 2014 lord63.
    :license: MIT, see LICENSE for more details.
"""

__title__ = "kindlepush"
__version__ = "0.2.0"
__author__ = "BlahGeek"
__maintainer__ = "lord63"
__license__ = "MIT"
__copyright__ = "Copyright 2014 BlahGeek 2014 lord63"

import logging
import sqlite3
import sys
import re
import os
import json
import HTMLParser

import requests
from bs4 import BeautifulSoup
from terminal import Command


# Set default encoding from ascii to utf-8
reload(sys)
if hasattr(sys, 'setdefaultencoding'):
    sys.setdefaultencoding('utf-8')

# User-agent is necessary, or it'll occur `KeyError: 'data' `.
GLOBAL_SESSION = requests.Session()
GLOBAL_SESSION.headers.update({'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux \
                               x86_64; rv:28.0) Gecko/20100101 Firefox/28.0'})


def unescape(raw_title):
    """unescape html character, to support chinese in title"""
    html_parser = HTMLParser.HTMLParser()
    if re.search(r'&.*;', raw_title):
        return html_parser.unescape(raw_title)
    else:
        return raw_title


def get_hidden_form_data(page):
    '''Get hidden form data from the page after you click the signin button

    Those data are needed to post to sign in.'''
    hidden_form_data = dict()
    input_tags = page.findAll('input', type='hidden')
    for tag in input_tags:
        hidden_form_data[tag['name']] = tag['value']
    return hidden_form_data


def login(config):
    '''To login in to amazom.com

    The long url is the 'sign in' botton, the short is the 'sign in our
    secure sever' botton.'''

    print 'Login...'
    login_url = ('https://www.amazon.com/ap/signin?_encoding=UTF8&openid.assoc'
                 '_handle=usflex&openid.claimed_id=http%3A%2F%2Fspecs.openid.n'
                 'et%2Fauth%2F2.0%2Fidentifier_select&openid.identity=http%3A%'
                 '2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openi'
                 'd.mode=checkid_setup&openid.ns=http%3A%2F%2Fspecs.openid.net'
                 '%2Fauth%2F2.0&openid.ns.pape=http%3A%2F%2Fspecs.openid.net%2'
                 'Fextensions%2Fpape%2F1.0&openid.pape.max_auth_age=0&openid.r'
                 'eturn_to=https%3A%2F%2Fwww.amazon.com%2Fgp%2Fyourstore%2Fhom'
                 'e%3Fie%3DUTF8%26ref_%3Dgno_signin')
    login_post_url = 'https://www.amazon.com/ap/signin'
    login_page = BeautifulSoup(GLOBAL_SESSION.get(login_url).text)
    data = get_hidden_form_data(login_page)
    data.update({'email': config['email'], 'password': config['password']})
    GLOBAL_SESSION.post(login_post_url, data)


def get_contents(config):
    """get all the information of docs in our kindle library"""
    url = ('https://www.amazon.com/gp/digital/fiona/manage/features/'
           'order-history/ajax/queryPdocs.html')
    request = GLOBAL_SESSION.post(url, {'offset': 0, 'count': config['count'],
                                        'contentType': 'Personal Document',
                                        'queryToken': 0, 'isAjax': 1})
    return [
        {'category': item['category'], 'contentName': item['asin'],
         'title': item['title']} for item in request.json()['data']['items']
        ]


def get_device_id():
    """automatically get the device id used when sending your docs"""
    request = GLOBAL_SESSION.post(
        'https://www.amazon.com/mn/dcw/myx/ajax-activity',
        data={'data': '{"param": {"GetDevices": {}}}'})
    device_id = request.json()['GetDevices']['devices'][0]['deviceAccountId']
    return device_id


def get_pending_deliveries():
    """get pending deliveries"""
    request = GLOBAL_SESSION.post(
        'https://www.amazon.com/mn/dcw/myx/ajax-activity',
        data={'data': '{"param": {"GetTodo": {}}}'})
    titles = [unescape(item['title']) for item in
              request.json()['GetTodo']['todoItems']]
    return titles


def deliver_content(content):
    """the function to deliver one doc"""
    url = ('https://www.amazon.com/gp/digital/fiona/content-download/'
           'fiona-ajax.html/ref=kinw_myk_ro_send')
    content.update({'isAjax': '1', 'deviceID': get_device_id()})
    request = GLOBAL_SESSION.post(url, content)
    # Whether successfully delivered it or not
    assert request.json()['data'] == 1


def deliver_all(contents, database):
    '''Deliver all the contents which haven't been delivered before

    Once delivered it, save it's asin in database.'''

    print 'Delivering'
    cursor = database.cursor()

    def content_in_database(content):
        """check whether a doc has been delivered or not"""
        try:
            cursor.execute('select * from content where name = "%s"' % content)
        except sqlite3.OperationalError:
            cursor.execute('create table content (name text)')
            return False
        else:
            return False if cursor.fetchone() is None else True

    contents = filter(lambda x: not content_in_database(x['contentName']),
                      contents)
    if not contents:
        sys.exit('All the docs in your library have been delivered'
                 'to your kindle')
    for content in contents:
        try:
            deliver_content(content)
            print 'delivering ' + unescape(content['title'])
        except Exception, error:
            logging.error(repr(error))
            print repr(error)
        else:
            cursor.execute('insert into content values ("%s")' %
                           content['contentName'])
            print 'Done. Save to database.'
            logging.info('delivered ' + unescape(content['title']))
    database.commit()


def main():
    command = Command('kindlepush',
                      'automatically deliver your docs to your kindle',
                      version=__version__)
    command.option('-c, --count [count]', 'the count of the docs to deliver')

    try:
        with open(os.path.join(sys.path[0], 'kindlepush_config.json')) as f:
            config = json.load(f)
    except IOError:
        sys.exit("Check your config file in {0}".format(sys.path[0]))

    @command.action
    def read(number=config['number']):
        """
        read the log file

        :param number: the count of days
        :option number: -n, --number [number]
        """
        file_path = os.path.join(config['directory'], 'kindlepush.log')
        os.system('tail -n {0} {1}'.format(number, file_path))

    @command.action
    def pending():
        """
        get pending deliveries
        """
        login(config)
        titles = get_pending_deliveries()
        print "Pending Deliveries:"
        for title in titles:
            print '\t{0}'.format(title)

    command.parse()

    logging.basicConfig(
        filename=os.path.join(config['directory'], 'kindlepush.log'),
        level='INFO',
        format='%(asctime)s [%(levelname)s] %(message)s')
    # Disable unwanted log message from the requests library
    requests_log = logging.getLogger("requests")
    requests_log.setLevel(logging.WARNING)

    # Connect to the database lying the directory specified in the config
    database = sqlite3.connect(os.path.join(config['directory'],
                                            'kindlepush.db'))

    # overwrite the value of count if user has specified it
    if command.count:
        config['count'] = command.count

    if not 'read' and 'pending' in sys.argv:
        try:
            login(config)
            deliver_all(get_contents(config), database)
        except KeyError:
            print 'KeyError, check your config file please.'
        except requests.exceptions.ConnectionError:
            print 'Check your network please.'

if __name__ == '__main__':
    main()
